### ISSUE TITLE
Something is wrong with the comparison

### ISSUE THREAD

**Issue by user1823** [open-spaced-repetition/fsrs-vs-sm17#25]
FSRS is indeed a good SRS model but, I can't believe that it would be better than SM-17. However, the results of the comparison made in this repo suggest that FSRS is better than SM-17.
<img width="336" alt="image" src="https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/92206575/8261c9de-41a1-44ba-8372-a519bd661b94">

This makes me think that there is something wrong with the comparison. However, I am unable to come up with reasonable causes for the poor performance of SM-17 against FSRS in this comparison.

**Comment by L-M-Sherlock** [comment]
> FSRS is indeed a good SRS model but, I can't believe that it would be better than SM-17.

What is your belief based on? If you find out the problem in this comparison, I would be glad to accept that and correct the result.

**Comment by L-M-Sherlock** [comment]
By the way, I have updated the result just now. Sometime SM-17 would predict 100% recall rate, which would induce a huge log loss. I have capped it in the latest code.

**Comment by L-M-Sherlock** [comment]
> Wozniak has done much research on spaced repetition algorithms over a course of several years and he continues to improve his model till now.

Yeah. I respect him. But I also have read most texts written by him and improved his theory with the data collected in China since five years ago: https://medium.com/@JarrettYe/how-did-i-publish-a-paper-in-acmkdd-as-an-undergraduate-c0199baddf31

By the way, I also sent emails to him to discuss the algorithm. But he always focuses on his education movement.

<img width="507" alt="image" src="https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/32575846/37edb65f-799a-44df-8722-475c876c1d54">

**Comment by Expertium** [comment]
I'm surprised as well. [The description of SM-17](https://supermemo.guru/wiki/Algorithm_SM-17) gives me the impression that it's more clever and sophisticated than FSRS, especially the "Three measures of retrievability" (grade-derived, R-matrix-derived and theoretical aka from a formula) and the way S is calculated.
![image](https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/83031600/ec211b95-8ed6-4434-8b0e-3f2721149127)

Unrelated, but Sherlock, I believe we stopped working on the R-matrix right when it started to show promising results. IIRC, it improved RMSE by 15% the last time we tried it, and it had a big flaw in grouping: reviews with very similar interval lengths would fall into different groups. There is no reason to put delta_t=365 and delta_t=366 into two distinct groups. And yet even with that flaw, the R-matrix helped to decrease RMSE. I'm confident that with proper grouping, RMSE can be improved further.

**Comment by user1823** [comment]
@Expertium, when you have the time, can you analyze the code used for comparison of the three models?

I think that there should be something that is making the comparison unfair. Unfortunately, I can't contribute much to the analysis because evaluation of the model is the part that I am least familiar with.

**Comment by Expertium** [comment]
It could be experience replaym I really doubt that SuperMemo uses it. Could also be the fact that FSRS is using a more modern gradient descent method ([Adam](https://pytorch.org/docs/stable/generated/torch.optim.Adam.html)), which has some heuristics to improve convergence and make the end result less dependent on learning rate.

**Comment by user1823** [comment]
@Expertium, good points, but these are the **strengths** of FSRS and not something that makes the comparison unfair.

The goal of this issue is to find and fix the **flaws** in the comparison process (if any).

One hypothetical example could be that the RMSE is calculated by different methods or by using different data for the two models.

Another hypothetical example could be that FSRS is using future data to train the model which is not available to SuperMemo. This problem was fixed by Sherlock when he started using online learning for FSRS.

**Comment by Expertium** [comment]
Then I don't know, I'm not that good at coding. You should ask @L-M-Sherlock

**Comment by user1823** [comment]
Well, my previous comment sounded like the goal is _just_ to find the coding errors. But, it is not.

We also need to find factors that are not bugs _per se_ but provide _undue_ advantage to FSRS in this comparison.

The best example that is currently coming to my mind is that FSRS was previously using future data to train the model which is not available to SuperMemo.
Stated another way, this means that to predict the memory states after a review, FSRS was using the data that was not even generated at the time of the review. This is clearly an undue advantage and this would have made the results greatly flawed.

Another example (which is not applicable in this comparison) is that the sizes of the bins to calculate the RMSE may be different for FSRS and SM.

As a counter example, the experience replay is not taken as an _undue_ advantage to FSRS because the data that FSRS is using for this purpose has already been generated and it is SuperMemo's shortcoming if SM-17 doesn't use it.

There may be other examples of such `undue` advantages. I can't think about most of them because I am not well versed with the process of evaluation.
